src/misc/io.c

```diff
@@ -87,6 +89,49 @@ void io_register_output(const char* name)
 	io_register(name, true);
 }
 
+void io_unregister(const char* name)
+{
+	struct iofile_s* iop = iofiles;
+	struct iofile_s* iop_old = iofiles;
+	struct iofile_s* tmp = NULL;
+	
+	while (NULL != iop) {
+		if (0 == strcmp(name, iop->name)) {
+			tmp = iop->prev;
+			XFREE(iop->name);
+			xfree(iop); // Note the lowercase, GODDAMMIT!
```

Comment:
iop has an indeterminate value after freeing, so strictly speaking you can not compare it.

Reply:
Nevermind, I just rewrote the list processing in idiomatic C to make it simpler. Only slightly tested.

Reply:
The value of the pointer itself becomes indeterminate after free. Actually, not only this pointer but all pointers which point to the object you have freed.

Reply:
But the `free` function does nothing to the value of the pointer passed as an argument, or am I missing something ?

The call to the `XFREE()`only zeros-out the `iop->name` pointer and not `iop`itself. Then the call to `xfree(iop)` releases the memory located at the location pointed to by `iop` and nothing else. Agreed, the location pointed to by `iop`is now invalid but `iop` itself still exists as a variable holding an address. The comparison `iofiles == iop` should therefore not pose a problem, since I am only looking at the value of the pointer themselves, which are still the same as before the two calls to `xfree` as neither `iop` nor `iofiles` have been modified until this point.

When developing this code, I ran some extensive tests using `valgrind` and did not detect any memory leaks nor accesses to invalid memory.

Reply:
No, the value of the pointer is then indeterminate. A compiler may use this for optimization and make the comparison be always false.